extensions/amp-auto-ads/0.1/placement.js

```diff
@@ -0,0 +1,224 @@
+    return resourcesForDoc(this.adElement_).attemptChangeSize(
+        this.adElement_, TARGET_AD_HEIGHT_PX, undefined).then(() => {
+          this.state_ = PlacementState.PLACED;
+          return this.state_;
+        }).catch(() => {
+          this.state_ = PlacementState.RESIZE_FAILED;
+          return this.state_;
+        });
+  }
+
+  /**
+   * @param {string} type
+   * @param {!Array<!DataAttributeDef>} dataAttributes
+   * @return {!Element}
+   * @private
+   */
+  createAdElement_(type, dataAttributes) {
+    const adElement = this.win_.document.createElement('amp-ad');
+    adElement.setAttribute('type', type);
+    adElement.setAttribute('layout', 'responsive');
+    adElement.setAttribute('width', '320');
+    adElement.setAttribute('height', '0');
+
+    for (let i = 0; i < dataAttributes.length; ++i) {
+      adElement.setAttribute(
+          'data-' + dataAttributes[i].name, dataAttributes[i].value);
+    }
+
+    return adElement;
+  }
+}
+
+/**
+ * @param {!Window} win
+ * @param {!JSONType} configObj
+ * @return {!Array<!Placement>}
+ */
+export function getPlacementsFromConfigObj(win, configObj) {
+  const placementObjs = configObj['placements'];
+  if (!placementObjs) {
+    dev().warn(TAG, 'No placements in config');
+    return [];
+  }
+  const placements = [];
+  for (let i = 0; i < placementObjs.length; ++i) {
+    const placement = getPlacementFromObject(win, placementObjs[i]);
+    if (placement) {
+      placements.push(placement);
+    }
+  }
+  return placements;
+}
+
+/**
+ * Validates that the placementObj represents a valid placement and if so
+ * constructs and returns a Placement object for it.
+ * @param {!Window} win
+ * @param {!Object} placementObj
+ * @return {?Placement}
+ */
+function getPlacementFromObject(win, placementObj) {
+  const anchor = placementObj['anchor'];
+  if (!anchor) {
+    dev().warn(TAG, 'No anchor in placement');
+    return null;
+  }
+  const anchorElement = getAnchorElement(win, anchor);
+  if (!anchorElement) {
+    dev().warn(TAG, 'No anchor element found');
+    return null;
+  }
+  const position = getPosition(placementObj['pos']);
+  if (!position) {
+    return null;
+  }
+  if ((position == Position.BEFORE || position == Position.AFTER) &&
+      !anchorElement.parentNode) {
+    dev().warn(TAG, 'Parentless anchor with BEFORE/AFTER position.');
+    return null;
+  }
+  return new Placement(win, anchorElement, position, placementObj);
+}
+
+/**
+ * @param {number} positionNum
+ * @return {?Position}
+ */
+function getPosition(positionNum) {
+  switch (positionNum) {
+    case Position.BEFORE:
+      return Position.BEFORE;
+    case Position.FIRST_CHILD:
+      return Position.FIRST_CHILD;
+    case Position.LAST_CHILD:
+      return Position.LAST_CHILD;
+    case Position.AFTER:
+      return Position.AFTER;
+    default:
+      dev().warn(TAG, 'Invalid position: ' + positionNum);
+      return null;
```

Comment:
oh, I had a bad example. Your snippet is the right version. Do you want to switch to your new code?